File: daedalus/core/internal/policy/policy.go
// Package policy 从 shared/policy.toml 读取 Daedalus 能力策略的单一事实源。
//
// 设计(计划 todo 12 / 决策 12):shell/fs 白名单、路径规则、净化环境与
// 超时不再以"散落的硬编码"为权威,而是集中定义于
// daedalus/files/system/opt/daedalus/shared/policy.toml(镜像内
// /opt/daedalus/shared/policy.toml),由 Go 能力服务器在启动时读取并经
// shellpolicy.WithPolicy / pathguard.WithAllowedDirs 注入。
//
// 路径解析优先级:
//  1. DAEDALUS_POLICY_PATH 环境变量(显式指向,文件损坏/缺失一律报错,
//     绝不静默吞掉——这是测试与 systemd drop-in 的注入口);
//  2. 生产路径 /opt/daedalus/shared/policy.toml;
//  3. 开发态回溯:自当前工作目录逐级上溯,寻找仓库内
//     daedalus/files/system/opt/daedalus/shared/policy.toml。
//
// 三处皆无 → ErrNotFound(调用方经 LoadOrDefault 回退 Default(),
// 保证"无 policy 服务器也能启动");显式指向的 TOML 损坏或字段缺失 →
// 加载失败报错(拒绝启动,fail-closed)。
//
// ALLOW_COMMANDS 环境变量维持 REPLACE 语义(与 shellpolicy.
// ResolveAllowCommands 现状一致):存在且非空时逗号分隔集合整体替换
// 策略中的 allowed_commands,而非取并集。
package policy

import (
	"errors"
	"fmt"
	"os"
	"path/filepath"
	"sort"
	"strings"

	"github.com/BurntSushi/toml"
)

// 环境变量名与候选路径常量。
const (
	// EnvPolicyPath 是策略文件的显式注入环境变量(测试/部署用)。
	EnvPolicyPath = "DAEDALUS_POLICY_PATH"
	// EnvAllowCommands 是命令白名单的整体替换环境变量(REPLACE 语义)。
	EnvAllowCommands = "ALLOW_COMMANDS"
	// ProductionPath 是镜像内的生产策略路径。
	ProductionPath = "/opt/daedalus/shared/policy.toml"
	// DevRelPath 是仓库内策略文件相对仓库根的路径(开发态回溯用,
	// 导出以便测试与构建脚本引用)。
	DevRelPath = "daedalus/files/system/opt/daedalus/shared/policy.toml"
)

// ErrNotFound 表示按解析优先级未找到任何策略文件。
// 与"文件存在但损坏/字段缺失"(真解析错误)严格区分:前者可回退
// Default(),后者必须拒绝启动。
var ErrNotFound = errors.New("policy: 未找到 policy.toml(生产路径与仓库回溯均未命中)")

// Shell 对应 TOML [shell] 表:shell 能力服务器的全部强制策略值。
type Shell struct {
	// AllowedCommands 是命令基名白名单(对应 allowed_commands)。
	AllowedCommands []string `toml:"allowed_commands"`
	// BinaryDirs 是允许的二进制所在目录(对应 binary_dirs)。
	BinaryDirs []string `toml:"binary_dirs"`
	// AllowedPathPrefixes 是路径参数允许的前缀(对应 allowed_path_prefixes)。
	AllowedPathPrefixes []string `toml:"allowed_path_prefixes"`
	// BlockedPaths 是显式禁止的敏感路径(对应 blocked_paths)。
	BlockedPaths []string `toml:"blocked_paths"`
	// CleanEnv 是子进程净化环境(对应 clean_env 内嵌表)。
	CleanEnv map[string]string `toml:"clean_env"`
	// TimeoutMs 是执行超时毫秒数(对应 timeout_ms)。
	TimeoutMs int64 `toml:"timeout_ms"`
}

// FS 对应 TOML [fs] 表:文件系统能力服务器的目录白名单。
type FS struct {
	// AllowedDirs 是可访问目录前缀(对应 allowed_dirs)。
	AllowedDirs []string `toml:"allowed_dirs"`
}

// Audit 对应 TOML [audit] 表:审计日志落点。
type Audit struct {
	// LogPath 是哈希链审计日志路径(对应 log_path)。
	LogPath string `toml:"log_path"`
}

// Policy 是 policy.toml 的完整解析结果,字段与 TOML 一一对应。
type Policy struct {
	Shell Shell `toml:"shell"`
	FS    FS    `toml:"fs"`
	Audit Audit `toml:"audit"`
}

// ResolvePath 按文档优先级解析策略文件路径。
// 未命中任何候选时返回包装 ErrNotFound 的错误;DAEDALUS_POLICY_PATH
// 指向的文件不存在**不算**未命中(由 Load 如实报读盘错误),
// 因为显式指向被静默降级会掩盖部署错误。
func ResolvePath() (string, error) {
	if p := os.Getenv(EnvPolicyPath); p != "" {
		return p, nil
	}
	if st, err := os.Stat(ProductionPath); err == nil && !st.IsDir() {
		return ProductionPath, nil
	}
	if wd, err := os.Getwd(); err == nil {
		for dir := wd; ; {
			cand := filepath.Join(dir, DevRelPath)
			if st, err := os.Stat(cand); err == nil && !st.IsDir() {
				return cand, nil
			}
			parent := filepath.Dir(dir)
			if parent == dir {
				break
			}
			dir = parent
		}
	}
	return "", ErrNotFound
}

// Load 解析指定路径的策略文件;path 为空串时先经 ResolvePath 走优先级链。
// 报错条件(fail-closed):文件读不出、TOML 语法损坏、存在未知键
// (拦截拼写错误,杜绝"以为改了其实没改")、必需字段缺失或为空。
func Load(path string) (*Policy, error) {
	if path == "" {
		var err error
		if path, err = ResolvePath(); err != nil {
			return nil, err
		}
	}
	data, err := os.ReadFile(path)
	if err != nil {
		return nil, fmt.Errorf("policy: 读取 %s 失败: %w", path, err)
	}
	var p Policy
	meta, err := toml.Decode(string(data), &p)
	if err != nil {
		return nil, fmt.Errorf("policy: 解析 %s 失败(TOML 损坏): %w", path, err)
	}
	if undecoded := meta.Undecoded(); len(undecoded) > 0 {
		keys := make([]string, 0, len(undecoded))
		for _, k := range undecoded {
			keys = append(keys, k.String())
		}
		sort.Strings(keys)
		return nil, fmt.Errorf("policy: %s 含未知键(fail-closed,拒绝拼写错误/第二事实源): %s", path, strings.Join(keys, ", "))
	}
	if err := p.validate(); err != nil {
		return nil, fmt.Errorf("policy: %s 校验失败: %w", path, err)
	}
	return &p, nil
}

// LoadOrDefault 是服务器启动入口:策略文件整体缺失时回退 Default()
// (保证无 policy 也能启动);文件存在但损坏/字段缺失时原样报错
// (调用方必须拒绝启动)。
func LoadOrDefault() (*Policy, error) {
	p, err := Load("")
	if errors.Is(err, ErrNotFound) {
		return Default(), nil
	}
	return p, err
}

// validate 钉死全部必需字段:任何缺段、缺键、空列表、空字符串、
// 非正超时都视为损坏策略。空列表一律拒绝是刻意的 fail-closed
// 设计——配置事故(误删整行)不得静默放宽或瘫痪安全边界。
func (p *Policy) validate() error {
	var missing []string
	requireList := func(name string, v []string) {
		if len(v) == 0 {
			missing = append(missing, name)
		}
	}
	requireList("shell.allowed_commands", p.Shell.AllowedCommands)
	requireList("shell.binary_dirs", p.Shell.BinaryDirs)
	requireList("shell.allowed_path_prefixes", p.Shell.AllowedPathPrefixes)
	requireList("shell.blocked_paths", p.Shell.BlockedPaths)
	requireList("fs.allowed_dirs", p.FS.AllowedDirs)

	if len(p.Shell.CleanEnv) == 0 {
		missing = append(missing, "shell.clean_env")
	} else if _, ok := p.Shell.CleanEnv["PATH"]; !ok {
		// PATH 缺失会让 argv 直发的裸命令名永远查不到可执行文件,
		// 属配置事故,必须在启动时拒绝而不是运行期全量失败。
		missing = append(missing, "shell.clean_env.PATH")
	}
	if p.Shell.TimeoutMs <= 0 {
		missing = append(missing, "shell.timeout_ms")
	}
	if p.Audit.LogPath == "" {
		missing = append(missing, "audit.log_path")
	}

	if len(missing) > 0 {
		sort.Strings(missing)
		return fmt.Errorf("缺失或为空的必需字段: %s", strings.Join(missing, ", "))
	}
	return nil
}

// AllocCommands 计算 shell 服务器的生效命令白名单:
// 读取 ALLOW_COMMANDS 环境变量并套用 REPLACE 语义(见 AllowedCommands)。
func AllocCommands(p *Policy) map[string]struct{} {
	return AllowedCommands(p, os.Getenv(EnvAllowCommands))
}

// AllowedCommands 是 AllocCommands 的纯函数形态(envValue 由调用方给出,
// 便于测试)。语义与 shellpolicy.ResolveAllowCommands 逐字对齐:
// envValue 非空 → 逗号分隔、逐项 trim、丢弃空项后的集合**整体替换**
// 策略白名单(非并集);envValue 为空 → 返回策略 allowed_commands 的集合副本。
func AllowedCommands(p *Policy, envValue string) map[string]struct{} {
	if envValue == "" {
		return commandSet(p.Shell.AllowedCommands)
	}
	allow := make(map[string]struct{})
	for _, c := range strings.Split(envValue, ",") {
		if trimmed := strings.TrimSpace(c); trimmed != "" {
			allow[trimmed] = struct{}{}
		}
	}
	return allow
}

// commandSet 把字符串切片转为集合(副本,防止调用方改动策略内部状态)。
func commandSet(cmds []string) map[string]struct{} {
	set := make(map[string]struct{}, len(cmds))
	for _, c := range cmds {
		set[c] = struct{}{}
	}
	return set
}

// Default 返回内嵌的出厂默认策略,值与 internal/shellpolicy、
// internal/pathguard 的既有硬编码常量逐项一致(15 命令 / 4 bin 目录 /
// 9 前缀 / 5 blocked / CLEAN_ENV / 30000ms / fs 3 目录 / 审计路径),
// 由 policy 包测试跨包比对钉死,防止双源漂移。
//
// 返回的是全新构造的实例(集合/映射均为独立副本),调用方可安全改动。
func Default() *Policy {
	return &Policy{
		Shell: Shell{
			AllowedCommands: []string{
				"df", "ls", "cat", "pwd", "uname", "free", "ps", "uptime",
				"whoami", "ip", "arch", "hostname", "date", "ping", "systemctl",
			},
			BinaryDirs: []string{"/usr/bin", "/bin", "/usr/sbin", "/sbin"},
			AllowedPathPrefixes: []string{
				"/home", "/var/log", "/tmp", "/proc", "/sys",
				"/etc/os-release", "/usr/lib/os-release", "/etc/fedora-release", "/etc/almalinux-release",
			},
			BlockedPaths: []string{
				"/etc/shadow", "/etc/gshadow", "/etc/sudoers", "/etc/sudoers.d", "/root",
			},
			CleanEnv: map[string]string{
				"PATH": "/usr/bin:/bin:/usr/sbin:/sbin",
				"LANG": "C.UTF-8",
			},
			TimeoutMs: 30000,
		},
		FS: FS{
			AllowedDirs: []string{"/home", "/var/log", "/tmp"},
		},
		Audit: Audit{
			LogPath: "/var/log/daedalus/audit.jsonl",
		},
	}
}

File: daedalus/files/system/opt/daedalus/plugins/daedalus.copilot/policy.ts
/**
 * Daedalus OS Copilot 策略与验证。
 *
 * 实现严格的提议验证、LLM 输出解析以及确定性的系统提示词生成。
 *
 * ★ 冻结副本声明 ★
 * 下方的白名单常量与网关校验器（validateCommand / validateArg /
 * validatePath / isPathLike）是 Go 侧 `daedalus/core/internal/shellpolicy`
 * 的**保持一致的冻结副本**（原始来源为已退役的 Deno shell 服务器实现）。
 * Copilot 在进程内做与 `daedalus-shell` Go 二进制完全相同的策略校验，
 * 保证“提议即执行”路径零策略偏离。
 * **修改 Go 侧 internal/shellpolicy 时必须同步修改本文件。**
 */

// 显式允许的诊断/只读命令（与 internal/shellpolicy 默认白名单一致的冻结副本，15 项）
export const DEFAULT_ALLOW_COMMANDS = new Set([
  "df",
  "ls",
  "cat",
  "pwd",
  "uname",
  "free",
  "ps",
  "uptime",
  "whoami",
  "ip",
  "arch",
  "hostname",
  "date",
  "ping",
  "systemctl",
]);

// 允许通过环境变量覆盖或扩展允许的命令
const envCommands = Deno.env.get("ALLOW_COMMANDS");
export const ALLOW_COMMANDS: Set<string> = envCommands
  ? new Set(envCommands.split(",").map((c) => c.trim()).filter((c) => c.length > 0))
  : DEFAULT_ALLOW_COMMANDS;

// 路径类参数允许的路径前缀（与 internal/shellpolicy 一致的冻结副本，9 项）
export const ALLOWED_PATH_PREFIXES = [
  "/home",
  "/var/log",
  "/tmp",
  "/proc",
  "/sys",
  "/etc/os-release",
  "/usr/lib/os-release",
  "/etc/fedora-release",
  "/etc/almalinux-release",
];

// 显式禁止的敏感路径（与 internal/shellpolicy 一致的冻结副本，5 项）
export const BLOCKED_PATHS = [
  "/etc/shadow",
  "/etc/gshadow",
  "/etc/sudoers",
  "/etc/sudoers.d",
  "/root",
];

/**
 * 判断参数是否类似于文件系统路径。
 * （internal/shellpolicy 冻结副本）
 */
export function isPathLike(arg: string): boolean {
  if (arg.includes("\0")) {
    return true;
  }
  if (
    arg.startsWith("/") ||
    arg.includes("/") ||
    arg === "." ||
    arg === ".." ||
    arg.startsWith("..")
  ) {
    return true;
  }
  return false;
}

/**
 * 针对不存在路径的基础路径规范化工具。
 * （internal/shellpolicy 冻结副本）
 */
function normalizePath(path: string): string {
  const parts = path.split("/").filter((p) => p.length > 0 && p !== ".");
  const stack: string[] = [];
  for (const part of parts) {
    if (part === "..") {
      stack.pop();
    } else {
      stack.push(part);
    }
  }
  return "/" + stack.join("/");
}

/**
 * 规范化并验证路径参数。
 * 确保路径不触及受阻路径且保持在允许的目录范围内。
 * （internal/shellpolicy 冻结副本）
 */
export function validatePath(pathStr: string): string {
  if (typeof pathStr !== "string" || pathStr.length === 0) {
    throw new Error("Path must be a non-empty string.");
  }
  if (pathStr.includes("\0")) {
    throw new Error("Null bytes are not allowed in path arguments.");
  }

  // 若路径存在则解析规范 realpath，若不存在则进行规范化
  let resolved: string;
  try {
    resolved = Deno.realPathSync(pathStr);
  } catch {
    // 若路径在磁盘上尚不存在，则相对于 cwd 或根路径解析
    const absolute = pathStr.startsWith("/")
      ? pathStr
      : `${Deno.cwd()}/${pathStr}`;
    resolved = normalizePath(absolute);
  }

  // 检查显式禁止的路径
  for (const blocked of BLOCKED_PATHS) {
    const cleanBlocked = blocked.replace(/\/+$/, "");
    if (resolved === cleanBlocked || resolved.startsWith(cleanBlocked + "/")) {
      throw new Error(`Access to blocked path '${pathStr}' (${resolved}) is forbidden.`);
    }
  }

  // 检查允许的目录前缀
  let allowed = false;
  for (const prefix of ALLOWED_PATH_PREFIXES) {
    const cleanPrefix = prefix.replace(/\/+$/, "");
    if (resolved === cleanPrefix || resolved.startsWith(cleanPrefix + "/")) {
      allowed = true;
      break;
    }
  }

  if (!allowed) {
    throw new Error(
      `Path '${pathStr}' (resolved: ${resolved}) is outside allowed directories: ${ALLOWED_PATH_PREFIXES.join(", ")}`,
    );
  }

  return resolved;
}

/**
 * 验证单个参数是否包含空字节和嵌入路径。
 * （internal/shellpolicy 冻结副本）
 */
export function validateArg(arg: string): void {
  if (typeof arg !== "string") {
    throw new Error(`Argument must be a string, got ${typeof arg}`);
  }
  if (arg.includes("\0")) {
    throw new Error("Null bytes are not allowed in arguments.");
  }

  if (arg.includes("=") && (arg.startsWith("-") || arg.startsWith("--"))) {
    const eqIdx = arg.indexOf("=");
    const val = arg.slice(eqIdx + 1);
    if (isPathLike(val)) {
      validatePath(val);
    }
  } else if (isPathLike(arg)) {
    validatePath(arg);
  }
}

/**
 * 依据白名单验证命令名称，并确保无路径遍历行为。
 * （internal/shellpolicy 冻结副本）
 */
export function validateCommand(command: string): string {
  if (!command || typeof command !== "string") {
    throw new Error("Command must be a non-empty string.");
  }
  if (command.includes("\0")) {
    throw new Error("Null bytes are not allowed in command.");
  }

  const trimmed = command.trim();
  const lastSlash = trimmed.lastIndexOf("/");
  const cmdBase = lastSlash >= 0 ? trimmed.slice(lastSlash + 1) : trimmed;

  if (!ALLOW_COMMANDS.has(cmdBase)) {
    throw new Error(`Command '${command}' is not in ALLOW_COMMANDS allowlist.`);
  }

  if (trimmed.includes("/")) {
    let cmdDir = trimmed.slice(0, lastSlash);
    try {
      cmdDir = Deno.realPathSync(cmdDir);
    } catch {
      // 若 realPathSync 失败则保持 cmdDir 原样
    }
    const allowedBinDirs = new Set(["/usr/bin", "/bin", "/usr/sbin", "/sbin"]);
    if (!allowedBinDirs.has(cmdDir)) {
      throw new Error(`Command path '${command}' is not in a valid system bin directory.`);
    }
  }

  return cmdBase;
}

export interface CommandProposal {
  command: string;
  args: string[];
  explanation: string;
}

/**
 * 将 LLM 输出解析并严格验证为 CommandProposal。
 * 如果存在 markdown 代码块标记则予以去除，并强制检查数据结构格式。
 */
export function parseProposal(text: string): CommandProposal {
  if (typeof text !== "string" || text.trim().length === 0) {
    throw new Error("LLM output schema validation failed: output must be non-empty text");
  }

  // 如果存在 markdown 代码块标记则予以去除（例如 ```json ... ``` 或 ``` ... ```）
  const cleanedText = text.replace(/```json|```/g, "").trim();

  let parsed: unknown;
  try {
    parsed = JSON.parse(cleanedText);
  } catch (err: unknown) {
    const msg = err instanceof Error ? err.message : String(err);
    throw new Error(`LLM output schema validation failed: invalid JSON: ${msg}`);
  }

  if (!parsed || typeof parsed !== "object" || Array.isArray(parsed)) {
    throw new Error("LLM output schema validation failed: root must be a JSON object");
  }

  const obj = parsed as Record<string, unknown>;

  if (typeof obj.command !== "string" || obj.command.trim().length === 0) {
    throw new Error("LLM output schema validation failed: 'command' must be a non-empty string");
  }

  if (!Array.isArray(obj.args)) {
    throw new Error("LLM output schema validation failed: 'args' must be an array of strings");
  }

  for (let i = 0; i < obj.args.length; i++) {
    if (typeof obj.args[i] !== "string") {
      throw new Error(`LLM output schema validation failed: 'args[${i}]' must be a string`);
    }
  }

  if (typeof obj.explanation !== "string") {
    throw new Error("LLM output schema validation failed: 'explanation' must be a string");
  }

  return {
    command: obj.command.trim(),
    args: obj.args as string[],
    explanation: obj.explanation,
  };
}

/**
 * 构建用于 LLM 命令转换的不可变、确定性系统提示词。
 */
export function buildSystemPrompt(): string {
  const allowedCommandsList = Array.from(ALLOW_COMMANDS).sort().join(", ");
  const allowedPrefixesList = ALLOWED_PATH_PREFIXES.join(", ");
  const blockedPathsList = BLOCKED_PATHS.join(", ");

  return `You are Daedalus OS Copilot, an AI assistant built into the operating system.
Your sole job is to translate user natural language requests into a single safe diagnostic / inspection command proposal.

RULES AND CONSTRAINTS:
1. You MUST only propose commands from this strict allowlist (ALLOW_COMMANDS):
   Allowed commands: [${allowedCommandsList}]

2. PATH SAFETY RULES:
   - Path arguments must be absolute and within allowed prefixes (ALLOWED_PATH_PREFIXES): [${allowedPrefixesList}]
   - Explicitly forbidden sensitive paths (BLOCKED_PATHS): [${blockedPathsList}]
   - Path traversal (..) is strictly forbidden.
   - Relative paths without leading slash are forbidden.
   - Null bytes (\\0) are strictly forbidden.

3. OUTPUT FORMAT:
   You must output ONLY valid JSON matching this exact schema:
   {
     "command": "<command name from allowlist>",
     "args": ["<arg1>", "<arg2>", ...],
     "explanation": "<brief human-readable explanation of what this command does>"
   }

4. STRICT PROHIBITIONS:
   - Output ONLY the JSON object. Do NOT include any markdown explanations, commentary, or text outside the JSON.
   - NEVER suggest or use commands outside the allowlist (e.g., rm, bash, sh, sudo, dd, chmod, chown are strictly forbidden).
   - NEVER use shell pipelines (|), redirects (>), subshells ($()), or chaining (&&, ;, ||). Command arguments must be a structured list of strings.
   - If the user's intent cannot be achieved using the allowed commands, choose the safest allowed diagnostic command or explain the limitation in the explanation field while keeping the command allowed (e.g. command: "uptime", explanation: "Requested action not permitted by OS safety policy.").`;
}

/**
 * 依据 DaedalusShell 安全网关规则验证 CommandProposal。
 * 若任一验证器拒绝则抛出错误。
 */
export function validateProposal(proposal: CommandProposal): void {
  if (!proposal || typeof proposal !== "object") {
    throw new Error("Invalid proposal: proposal must be an object");
  }

  // 1. 根据 ALLOW_COMMANDS 和二进制路径验证命令
  validateCommand(proposal.command);

  // 2. 验证每个参数
  if (Array.isArray(proposal.args)) {
    for (const arg of proposal.args) {
      validateArg(arg);
    }
  }
}
